### DISCUSSION TITLE
Make kmonad execute on startup

### DISCUSSION THREAD

**Discussion by michaelmoreno** [kmonad/kmonad#689] [Q&A]
I currently have to manually execute
```bash
$ sudo kmonad .config/kmonad/asus-m16.kbd
$ xset r rate 200 30
```
every session to get my keyboard configured properly. I would like this to happen automatically on login / system startup.

I attempted to do this by adding it as a line in my `.xinitrc`
```bash
#!/bin/sh

userresources=$HOME/.Xresources
usermodmap=$HOME/.Xmodmap
sysresources=/etc/X11/xinit/.Xresources
sysmodmap=/etc/X11/xinit/.Xmodmap

# merge in defaults and keymaps

if [ -f $sysresources ]; then
    xrdb -merge $sysresources
fi

if [ -f $sysmodmap ]; then
    xmodmap $sysmodmap
fi

if [ -f "$userresources" ]; then
    xrdb -merge "$userresources"
fi

if [ -f "$usermodmap" ]; then
    xmodmap "$usermodmap"
fi

# start some nice programs

if [ -d /etc/X11/xinit/xinitrc.d ] ; then
 for f in /etc/X11/xinit/xinitrc.d/?*.sh ; do
  [ -x "$f" ] && . "$f"
 done
 unset f
fi

eval $(gnome-keyring-daemon --start)
export SSH_AUTH_SOCK

sudo kmonad .config/kmonad/asus-m16.kbd # <--- this line
xset r rate 200 30
xbindkeys
dunst &
# autokey-qt &
picom --experimental-backends &
nitrogen --restore &
exec qtile start
```
This results in a black screen on login, bricking my system until I go back and remove the line in a `tty`.
I've tried adding it at the end as well
```bash
exec qtile start
sudo kmonad .config/kmonad/asus-m16.kbd
```
which also results in a black screen.

What is the recommended way to achieve automatic activation?

**Comment by gerardbosch** [comment]
Hi! For Linux you can use systemd to start it for you. If you want to make it available only for your user, use this service unit:

https://github.com/gerardbosch/dotfiles-linux/blob/main/home/.config/systemd/user/kmonad%40.service

[Permalink (in case the above link breaks due repo changes) »](https://github.com/gerardbosch/dotfiles-linux/blob/66621f0998ce7984358abd9d4e05c5dd5696361e/home/.config/systemd/user/kmonad%40.service)

* Copy it under `~/.config/systemd/user/`
* I think you need to enable it the first time with something like `systemctl --user enable kmonad` (don't remember exact)
* Then it will be auto started on login (both graphical or TTY login). You can also manage it with:
  * `systemctl --user status kmonad@<your-config-file>`
  * `systemctl --user restart kmonad@<your-config-file>`
  * `systemctl --user stop kmonad@<your-config-file>`
  * `systemctl --user start kmonad@<your-config-file>`
  * `journalctl -f --user -u kmonad@<your-config-file>` to watch the logs

This is using systemd in `--user` mode (no root required).

**Reply by luroc** [reply]
This is one of the first search results for this topic, but the link is broken. Could you fix the link or just paste your service file here?

**Reply by gerardbosch** [reply]
Hello! I updated the link on the original message, now points to the right file again :)

Bear in mind that the file is adapted for my own setup, so take it as inspiration.

I have 2 keyboards (laptop built-in + external keyboard) and I have both intercepted by KMonad. That's why I have a systemd service that starts for each keyboard. I have this (I use [HomeSHick](https://github.com/andsens/homeshick) to manage my dotfiles, that's why those symlinks):

```
❯ s ~/.config/kmonad                 
Permissions Size User   Date Modified Name
lrwxrwxrwx     - gerard 29 jun  2025  external-iso-105-keys.kbd -> ../../.homesick/repos/dotfiles/home/.config/kmonad/external-iso-105-keys.kbd
lrwxrwxrwx     - gerard 29 jun  2025  laptop-ansi-104-keys.kbd -> ../../.homesick/repos/dotfiles/home/.config/kmonad/laptop-ansi-104-keys.kbd
```

**Comment by jokesper** [comment]
Depending on your need there is are also startup scripts under [startup/](https://github.com/kmonad/kmonad/tree/master/startup) or in [kmonad-contrib/scripts](https://github.com/kmonad/kmonad-contrib/tree/master/scripts).
